### ISSUE TITLE
Process magit-todos--scan-with-git-diff failed

### ISSUE THREAD

**Issue by Shinmera** [alphapapa/magit-todos#153]
Hi, I'm getting a very strange error on every magit operation I do, making it quite annoying. `*messages*` shows the following:

> error in process sentinel: magit-todos--async-when-done: Process "magit-todos--scan-with-git-diff" failed with exit code 125.  Output:"/usr/bin/nice: a command must be given with an adjustment
Try '/usr/bin/nice --help' for more information.
"

I have no idea what that relates to or where I should look in my environment to figure out what's going on. Any pointers?

**Comment by alphapapa** [comment]
When did it start happening?  What happens if you disable `magit-todos-nice`?

**Comment by Shinmera** [comment]
I'm not sure. It can't have been more than some days ago. It doesn't happen on my other machines, but I have very similar setups.

How would I disable that? All I have in my init related to magit-todos is:
```elisp
(magit-todos-mode)
(setq magit-todos-keywords-list (mapcar #'car hl-todo-keyword-faces))
```

**Comment by alphapapa** [comment]
`M-x customize-option RET magit-todos-nice RET`

**Comment by Shinmera** [comment]
Disabling nice seems to remove the errors.

As for why, I'm not sure, might just be leftover config from forever ago. I've had this set up for years and never touched it after the first time.

**Comment by alphapapa** [comment]
I'm guessing that disabling the use of `nice` merely obscures the real problem, because you're reporting an error from the branch diff scanner, and `nice` is still used for the main scanner, which doesn't seem to be signaling an error.

I'm guessing it might have been caused by this commit: https://github.com/alphapapa/magit-todos/commit/8da7a8e9008281999567a2fb39f2d13d102abfbb  But it probably only happens under certain conditions.  So please re-enable the use of `nice`, then cause the error again, and then check the value of `(magit-main-branch)` in that repository.  As well, please check the messages buffer for any other errors around the same time (if multiple errors happened, you might only have noticed the last one).

**Comment by Shinmera** [comment]
Okey, here's the log from messages:
```
You can run the command ‘magit-status’ with C-x g
error in process sentinel: magit-todos--async-when-done: magit-todos--async-when-done: process "magit-todos--scan-with-git-diff" failed with exit code 128.  Output:"fatal: invalid object name 'fatal'.
"
error in process sentinel: magit-todos--async-when-done: process "magit-todos--scan-with-git-diff" failed with exit code 128.  Output:"fatal: invalid object name 'fatal'.
"
"main"
```
The main there being the result of magit-main-branch.

Also I just noticed that I set ``hl-todo-keyword-faces`` earlier in my config, hence the customisation:
```elisp
(setq hl-todo-keyword-faces '(("KLUDGE" . "#d0bf8f")
                         ("HACK" . "#d0bf8f")
                         ("TODO" . "#D98C32")
                         ("FIXME" . "#EF681F")
                         ("WTF" . "#F8240E")))
```

**Comment by alphapapa** [comment]
Ok, please run this command in that repo (ensuring that the same branch is checked out as when the error happened): `git merge-base HEAD main`

> Also I just noticed that I set `hl-todo-keyword-faces` earlier in my config, hence the customisation:

Ok, FYI, if `magit-todos` is loaded after you do that, it should include that customization you've made, because the `defcustom` setter function will read that value.  And note as well that setting the value manually precludes the option `magit-todos-ignored-keywords` from taking effect, because it's accounted for in the same setter.

In other words, this is the classic admonition to not use `setq` on `defcustom` options.  It will work fine, until it doesn't, and then mysterious behavior will result.  In Emacs 29, however, you can use `setopt`, which runs the setter.

**Comment by Shinmera** [comment]
`git merge-base HEAD main` reports: `fcff5272894da3427d5f43edb7aee9692ee35f67`

**Comment by alphapapa** [comment]
What branch or commit do you have checked out when the error occurs?

**Comment by Shinmera** [comment]
Current HEAD, which is the commit I just posted.

**Comment by alphapapa** [comment]
I am unable to reproduce the problem.  Using Emacs 29.1:

1. Run `mkdir /tmp/test && cd /tmp/test && git clone https://github.com/shinmera/multiposter && cd multiposter`.
2.  Run `with-emacs.sh -e emacs-29.1 -i magit-todos -- -f magit-todos-mode -f magit-status`.

The magit-status buffer is shown and there are no errors in the messages buffer.  If I add a to-do item to one of the Lisp files and refresh the status buffer, the todo is shown, and still no errors.

**Comment by Shinmera** [comment]
Okey, the other machine was updated but had autoremovable packages. I guess there must have been interference from a previous version of something or another.

Many apologies for the noise, and thanks a lot for the help debugging this!

**Comment by alphapapa** [comment]
You're welcome.  Thanks.

**Comment by alphapapa** [comment]
@Shinmera  Well, the same error finally happened to me, when I was in an orphan branch in one of my projects' repos.  I just pushed a fix for it: https://github.com/alphapapa/magit-todos/commit/a61d6af2852c6a1492ca157307e8bcf4cbed9122  If you can, please let me know if it fixes the problem for you.  Or if you can't reproduce the problem first in order to test it, let me know if anything seems to break, and if not, I'll release it as a bug fix.  Thanks.

**Comment by accelbread** [comment]
Hit this during a rebase. The mentioned fix does not fix it for me.

**Comment by alphapapa** [comment]
@accelbread I appreciate your report, but I'm afraid I can't do anything without a reproducible test case.  There is too much variation between platforms and the various software versions to know what is going on.  If you can produce one, e.g. using [this comment](https://github.com/alphapapa/magit-todos/issues/153#issuecomment-1713779857) as an example, maybe it can be solved.  As well, the fact that it happened during a rebase could imply a transient problem caused by one scanner or the other seeing worktree files in an inconsistent state, or something like that; if the problem went away, then that may just have to be good enough for now.

**Comment by alphapapa** [comment]
I found this bug again and am pushing a fix for it in v1.7.1.

@accelbread Maybe the bug you're encountering is a different one or has a different cause.  If it happens to you again after upgrading to v1.7.1, please file a new report.
